### PR TITLE
feat: Added like feature (frontend + backend)

### LINKED ISSUE #305: [FEATURE] Add a Like or Upvote button for each project
### Description

It would be nice having a Like or upvote button for project, so user's of the site would be able like the project, also it will put a soft impact on visitors of the project.

### LINKED ISSUE #306: [FEATURE] add a upvote or like button - Backend
### Description

Add upvote functionality to the backend of project, count all the upvotes for project, and update count as it grows.

### PR DESCRIPTION
**Pull request by zvolcsey**
Hi, I created one branch that contains the frontend and backend parts of like-feature.

Should close #305 
Should close #306 

I do not solve this task yet:
Can you help?
- Rerender list of projects after login/logout.

### CODE REVIEW

**Comment by rohitdasu** [comment]
> Hi, I created one branch that contains the frontend and backend parts of like-feature.
> 
> Should close #305 Should close #306
> 
> I do not solve this task yet: Can you help?
> 
> * Rerender list of projects after login/logout.

could you explain the issue a bit? maybe u can share info in our discord channels

**Comment by zvolcsey** [comment]
Hi @rohitdasu thanks for offering your help. The issue was that when a user was logged in and liked a feature, and then they logged out, the page still showed that project as liked by that user. I managed to fix it now by invalidating and refetching the project data every time a user likes a project or logs out, but as a result, there is a significant delay between the user clicking on the like button and seeing a visual feedback of their action. I’m still working on a solution to make the user action seem instant in the client.

**Comment by rohitdasu** [comment]
> Hi @rohitdasu thanks for offering your help. The issue was that when a user was logged in and liked a feature, and then they logged out, the page still showed that project as liked by that user. I managed to fix it now by invalidating and refetching the project data every time a user likes a project or logs out, but as a result, there is a significant delay between the user clicking on the like button and seeing a visual feedback of their action. I’m still working on a solution to make the user action seem instant in the client.

thanks @zvolcsey for trying hard. please don't feel overwhelmed and take your time finding better solution. we are keeping this issue assigned to you only. 👍

**Comment by rohitdasu** [comment]
when i try to like a project it's showing me this - 
Screenshot 2023-01-22 at 9 01 52 PM

**Comment by zvolcsey** [comment]
@rohitdasu Thanks for the feedback! Working on it.

**Comment by zvolcsey** [comment]
Hi @rohitdasu, in my latest commits I made the following improvements:
- The user logging in or out or navigating away from the page refetches the entire project list to make sure that the like status is up to date.
- The user sees immediately the state change after liking/disliking a project, but the change is rolled back on the front end if the request returns an error.
- The like/unlike button is now disabled while there is a request in flight.
- On server side the like/unlike functions return a like and a project with author data.

The PR is now ready for review from my side 😃.

File: components/Projects/Project.tsx [hunk 1]
```diff
@@ -1,23 +1,95 @@
...
   author,
+  liked,
+  likesCount,
 }: ProjectProps) => {
+  const session = useSession();
+  const [likedState, setLiked] = useState(liked);
+  const [likesCountState, setLikesCount] = useState(likesCount);
+  const [likeBtnIsDisabled, setLikeBtnIsDisabled] = useState(false);
+
+  const isOwnedByUser = author === session.data?.user?.name;
```
[the remark this answers is not in the document]
Comment (zvolcsey) [reply components/Projects/Project.tsx:27]: The author of the project is equal to the logged User. The user cannot like her/his own project.

[the remark this answers is not in the document]
Comment (rohitdasu) [reply components/Projects/Project.tsx:27]: exactly, i understand that. but there is a chance that, author of the project and logged user is same. can it handle that situation?

[the remark this answers is not in the document]
Comment (zvolcsey) [reply components/Projects/Project.tsx:27]: Thanks for the feedback! Working on it.

[the remark this answers is not in the document]
Comment (zvolcsey) [reply components/Projects/Project.tsx:27]: Is it good, if I check the email?
- The email is unique.
- The getAllProject function (on the server) collects the name and the email in the author property.
- The `Project` component gets an author object.

author

[the remark this answers is not in the document]
Comment (zvolcsey) [reply components/Projects/Project.tsx:27]: `const isOwnedByUser = author.email === session.data?.user?.email;`

[the remark this answers is not in the document]
Comment (rohitdasu) [reply components/Projects/Project.tsx:27]: can we compare user ids?

[the remark this answers is not in the document]
Comment (zvolcsey) [reply components/Projects/Project.tsx:27]: I also thought about ids, but User ids are not currently part of the session. That's why I mentioned the email. Emails are currently part of the session and are unique. The solution with id requires more changes.

Let me know please which would be better!

The solution with id:
Returns the id of the project's author is simple. The getAllProject function (on the server) can collect the author`s id.
I read the nextAuth docs and I tried to add the User id to the session.
- `next-auth.d.ts` - I added the id to the Session type. (new file)
- `[...nextAuth].ts` - The User id add to the session in the callbacks property. In development works well. Currently I don`t know, how it works in production with callbacks property.
- `tsconfig.json` - I added the `next-auth.d.ts` file to the `tsconfig.json`.

next-auth
nextAuth
tsconfig

[the remark this answers is not in the document]
Comment (zvolcsey) [reply components/Projects/Project.tsx:27]: I used this links for the solution with id.
https://next-auth.js.org/getting-started/typescript#module-augmentation
https://stackoverflow.com/a/71721634

[the remark this answers is not in the document]
Comment (rohitdasu) [reply components/Projects/Project.tsx:27]: ok @zvolcsey 

if comparing IDs is more complex for now then we can use emails. 👍

**Comment by rupali-codes** [comment]
Hi, can we instead disable the button when user is not logged in? and add a `tooltip` when user tries to Like it? It would be better ig, rather than showing that error pop-up.

image

what are your thoughts @rohitdasu ?

**Comment by rohitdasu** [comment]
> Hi, can we instead disable the button when user is not logged in? and add a `tooltip` when user tries to Like it? It would be better ig, rather than showing that error pop-up.
> 
> 
> 
> image
> 
> 
> 
> what are your thoughts @rohitdasu ?

I believe disabling button is fine for now. Later we can add tooltip. Currently we don't have that component. We can add it in separate PR.

**Comment by zvolcsey** [comment]
@rohitdasu @rupali-codes Thanks for the feedback! Then, I compare emails and disable the like button if the session is unauthorized.

**Comment by rohitdasu** [comment]
> @rohitdasu @rupali-codes Thanks for the feedback! Then, I compare emails and disable the like button if the session is unauthorized.

Cool 👌🏼

**Comment by zvolcsey** [comment]
@rohitdasu Is cursor: not-allowed CSS class necessary if the like button is disabled or the cursor stays unchanged?

**Comment by rohitdasu** [comment]
> @rohitdasu Is cursor: not-allowed CSS class necessary if the like button is disabled or the cursor stays unchanged?

Cursor should remain same 👍

**Comment by rupali-codes** [comment]
> > @rohitdasu Is cursor: not-allowed CSS class necessary if the like button is disabled or the cursor stays unchanged?
> 
> Cursor should remain same 👍

ig we should disable cursor too?

**Comment by rohitdasu** [comment]
> > > @rohitdasu Is cursor: not-allowed CSS class necessary if the like button is disabled or the cursor stays unchanged?
> > 
> > 
> > Cursor should remain same 👍
> 
> ig we should disable cursor too?

how? you mean making it `not-allowed` cursor?

**Comment by rupali-codes** [comment]
> > > > @rohitdasu Is cursor: not-allowed CSS class necessary if the like button is disabled or the cursor stays unchanged?
> > > 
> > > 
> > > Cursor should remain same 👍
> > 
> > 
> > ig we should disable cursor too?
> 
> how? you mean making it `not-allowed` cursor?

yeah

**Comment by zvolcsey** [comment]
@rohitdasu @rupali-codes I fixed the following things:
- The `Project` component now compares emails.
- The likeBtn is disabled, if the user is not logged in. The cursor is currently default, not pointer. I saw the `Button` component and the disabled button has default cursor. If you are going to choose the not-allowed cursor, then I will fix it. For security reason stays the status check in the `handleClick` function. If the user changes the button status in dev tools, then gets a UI error, too.

**Review by rupali-codes — COMMENTED** [review]
LGTM. Just update the cursor to `not-allowed`

**Comment by zvolcsey** [comment]
> LGTM. Just update the cursor to `not-allowed`

The `Button` component overwrites the cursor class, if I add `not-allowed` class in the `Project` component to the button. I should update the `cursor-default` to `cursor-not-allowed` in the reusable `Button` component. Every button has this change. Is this not a problem?

button

**Comment by rohitdasu** [comment]
> > LGTM. Just update the cursor to `not-allowed`
> 
> The `Button` component overwrites the cursor class, if I add `not-allowed` class in the `Project` component to the button. I should update the `cursor-default` to `cursor-not-allowed` in the reusable `Button` component. Every button has this change. Is this not a problem?
> 
> button

i think let's change the component's cursor default. let's make it `cursor-not-allowed`.

**Comment by zvolcsey** [comment]
@rohitdasu @rupali-codes The disabled button has a `not-allowed` cursor now.

**Comment by rupali-codes** [comment]
> @rohitdasu @rupali-codes The disabled button has a `not-allowed` cursor now.

awesome, but it's still blinking on hover (when disabled) ;)

**Comment by rohitdasu** [comment]
> > @rohitdasu @rupali-codes The disabled button has a `not-allowed` cursor now.
> 
> awesome, but it's still blinking on hover (when disabled) ;)

that's fine in my opinion. we can deal it in different PR

**Comment by zvolcsey** [comment]
> > > @rohitdasu @rupali-codes The disabled button has a `not-allowed` cursor now.
> > 
> > 
> > awesome, but it's still blinking on hover (when disabled) ;)
> 
> that's fine in my opinion. we can deal it in different PR. I will create a seperate PR for some UI issues related to this.

Okay, I do not deal with this issue now. Is there any other issue with the like-feature that I need to fix?

**Comment by rohitdasu** [comment]
> > > > @rohitdasu @rupali-codes The disabled button has a `not-allowed` cursor now.
> > > 
> > > 
> > > awesome, but it's still blinking on hover (when disabled) ;)
> > 
> > 
> > that's fine in my opinion. we can deal it in different PR. I will create a seperate PR for some UI issues related to this.
> 
> Okay, I do not deal with this issue now. Is there any other issue with the like-feature that I need to fix?

there is a one issue, when we login or logout the projects get refreshing... i think its not a big issue but not a good user experience. if it's not fixable let me know.

**Comment by zvolcsey** [comment]
> > > > > @rohitdasu @rupali-codes The disabled button has a `not-allowed` cursor now.
> > > > 
> > > > 
> > > > awesome, but it's still blinking on hover (when disabled) ;)
> > > 
> > > 
> > > that's fine in my opinion. we can deal it in different PR. I will create a seperate PR for some UI issues related to this.
> > 
> > 
> > Okay, I do not deal with this issue now. Is there any other issue with the like-feature that I need to fix?
> 
> there is a one issue, when we login or logout the projects get refreshing... i think its not a big issue but not a good user experience. if it's not fixable let me know.

At first I tried with the solution without show skeleton, but it was unsuccessful. This solution probably take more time.  I had problems with the revalidation of the liked property. 

Then I chose this solution. The `ProjectsList` component has a local projects state. If the user login or logout, the setProjects function update the local state with an empty array. The mutate function returns the up to date projects. After that runs the filter function. Until the data is not valid, the user sees the skeleton.
I think that  we can think about a better user experience solution  in a separate issue.

`ProjectsList` component
auth

**Comment by rohitdasu** [comment]
i'm merging it in seperate branch for now - later after finding the proper fix of the current issue will merge it to main branch.

**Review by rohitdasu — APPROVED** [review]

**Pull request merged**